testing/e2e/pkg/testnet.go

```diff
@@ -0,0 +1,315 @@
+				if otherName == name {
+					continue
+				}
+				node.Peers = append(node.Peers, otherName)
+			}
+		}
+		// replace the peer names with the P2P address.
+		for idx, peer := range node.Peers {
+			node.Peers[idx] = testnet.Nodes[peer].AddressP2P(true)
+		}
+	}
+
+	// deterministically sort accounts in alphabetical order
+	accountNames := []string{}
+	for name := range manifest.Accounts {
+		accountNames = append(accountNames, name)
+	}
+	sort.Strings(accountNames)
+
+	for _, name := range accountNames {
+		accountManifest := manifest.Accounts[name]
+		if _, ok := testnet.Accounts[name]; ok {
+			return nil, fmt.Errorf("duplicate account name %s", name)
+		}
+		account := &Account{
+			Name:   name,
+			Tokens: accountManifest.Tokens,
+			Key:    keyGen.Generate(accountManifest.KeyType),
+		}
+		if account.Tokens == 0 {
+			account.Tokens = defaultAccountTokens
+		}
+		testnet.Accounts[name] = account
+	}
+
+	return testnet, testnet.Validate()
+}
+
+func (t *Testnet) Validate() (err error) {
+	if len(t.Accounts) == 0 {
+		return errors.New("at least one account is required")
+	}
+	if len(t.Nodes) == 0 {
+		return errors.New("at least one node is required")
+	}
+	validators := 0
+	for name, node := range t.Nodes {
+		if err := node.Validate(); err != nil {
+			return fmt.Errorf("invalid node %s: %w", name, err)
+		}
+		// must have at least one validator
+		if node.SelfDelegation > 0 {
+			validators++
+		}
+	}
+	if validators == 0 {
+		return errors.New("at least one node must a validator by having an associated account")
+	}
+	for _, account := range t.Accounts {
+		if err := account.Validate(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (t *Testnet) GetAllVersions() []string {
+	versions := make(map[string]struct{})
+	// filter duplicate version strings
+	for _, node := range t.Nodes {
+		for _, version := range node.Versions {
+			versions[version] = struct{}{}
+		}
+	}
+
+	// convert to list
+	versionsList := []string{}
+	for version := range versions {
+		versionsList = append(versionsList, version)
+	}
+	return versionsList
+}
+
+func (t *Testnet) NodesByStartHeight() []*Node {
+	nodes := make([]*Node, 0, len(t.Nodes))
+	for _, node := range t.Nodes {
+		nodes = append(nodes, node)
+	}
+	sort.Slice(nodes, func(i, j int) bool {
+		if nodes[i].StartHeight == nodes[j].StartHeight {
+			return nodes[i].Name < nodes[j].Name
+		}
+		return nodes[i].StartHeight < nodes[j].StartHeight
+	})
+	return nodes
+}
+
+// Address returns a P2P endpoint address for the node.
+func (n Node) AddressP2P(withID bool) string {
+	addr := fmt.Sprintf("%v:26656", n.IP.String())
+	if withID {
+		addr = fmt.Sprintf("%x@%v", n.NetworkKey.PubKey().Address().Bytes(), addr)
+	}
+	return addr
+}
+
+// Address returns an RPC endpoint address for the node.
+func (n Node) AddressRPC() string {
+	return fmt.Sprintf("%v:26657", n.IP.String())
+}
+
+func (n Node) IsValidator() bool {
+	return n.SelfDelegation != 0
```

Comment:
[optional] Since a non-zero `SelfDelegation` is the only requirement for being recognized as a validator, why not use a boolean type for `SelfDelegation` instead?

Reply:
Because we may want to set the self delegation to an actual number which will reflect the initial voting power of the validator